File: src/main/java/leetcode/ValidParentheses.java
package leetcode;

import java.util.Stack;

/**
 * Given a string containing just the characters '(', ')', '{', '}', '[' and
 * ']', determine if the input string is valid.
 *
 * An input string is valid if:
 *
 * Open brackets must be closed by the same type of brackets. Open brackets must
 * be closed in the correct order. Note that an empty string is also considered
 * valid.
 */
final class ValidParentheses {

    private ValidParentheses() {
        // prevent instantiation
    }

    /**
     * Return whether a String contains valid parenthesis.
     *
     * @param s the input String
     * @return if the input String is a valid input string
     */
    public static boolean isValid(final String s) {
        Stack<Character> myStack = new Stack<>();

        for (int i = 0; i < s.length(); i++) {

            char ch = s.charAt(i);
            if (ch == '(' || ch == '[' || ch == '{') {
                myStack.add(ch);
            } else if (ch == ')' || ch == ']' || ch == '}') {
                if (myStack.isEmpty()) {
                    return false;
                }
                switch (ch) {
                case ')':
                    if (myStack.peek() != '(') {
                        return false;
                    }
                    break;
                case ']':
                    if (myStack.peek() != '[') {
                        return false;
                    }
                    break;
                case '}':
                    if (myStack.peek() != '{') {
                        return false;
                    }
                    break;
                default:
                }
                myStack.pop();
            } else {
                return false;
            }
        }
        return myStack.isEmpty();
    }
}

File: python/src/valid_parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        """
        Given a string s containing just the characters '(', ')', '{', '}', '[' and ']', determine if the input
        string is valid.

        :param s: The input string.
        :type s: Str.
        :return: True if s is valid, and False otherwise.
        :rtype: Bool

        The time complexity is O(N) where N is the length of s. Faster than 90.82% of solutions.

        The space complexity is O(N) where N is the length of s. Less memory than 69.44% of solutions.
        """
        stack = []
        closeToOpen = {")": "(", "]": "[", "}": "{"}

        for c in s:
            if c in closeToOpen:
                if stack and stack[-1] == closeToOpen[c]:
                    stack.pop()
                else:
                    return False
            else:
                stack.append(c)

        return len(stack) == 0
